### PR TITLE
enhancement(http source): Add support for metrics to the HTTP sink and source

### PR DESCRIPTION
**Pull request by Xaeroxe**
This PR adds support for metrics to the HTTP sink and source. This support changes the text format, ~~so in order to maintain backwards compatibility with existing deployments this feature is **opt in** via the `metrics_plus_mode` boolean configuration option.~~

At reviewer request, metrics plus mode has been removed.

### CODE REVIEW

**Comment by Xaeroxe** [comment]
That timestamp format test doesn't pass on master either :frowning:

File: src/sinks/http.rs [hunk 1]
```diff
@@ -138,35 +141,54 @@ impl SinkConfig for HttpSinkConfig {
...
     }
 
     fn sink_type(&self) -> &'static str {
         "http"
     }
 }
 
+#[derive(Serialize)]
+#[serde(untagged)]
+enum EventOrLog {
```
Comment (jszwedko) [inline src/sinks/http.rs:154]: I'm not seeing why this enum is required. Could we not just encode `Event`?

File: src/sinks/http.rs [hunk 2]
```diff
@@ -336,6 +358,21 @@ mod tests {
         assert_eq!(output.message, "hello world".to_string());
     }
 
+    #[test]
+    fn http_encode_event_json_metrics_plus_mode() {
+        let encoding = EncodingConfig::from(Encoding::Ndjson);
+        let event = Event::from("hello world");
```
Comment (jszwedko) [inline src/sinks/http.rs:364]: I think this is just creating a `Event::Log` event. We probably want to test a `Event::Metric` here. You can create one directly with `Event::Metric(Metric{ ... })`.

File: src/sources/http.rs [hunk 3]
```diff
@@ -48,13 +51,16 @@ pub enum Encoding {
...
+        decode_body(body, self.encoding, self.metrics_plus_mode)
             .map(|events| add_headers(events, &self.headers, header_map))
             .map(|mut events| {
                 // Add source type
                 let key = log_schema().source_type_key();
                 for event in events.iter_mut() {
-                    event.as_mut_log().try_insert(key, Bytes::from("http"));
+                    match event {
+                        Event::Log(event) => event.try_insert(key, Bytes::from("http")),
+                        Event::Metric(_) => (),
```
Comment (jszwedko) [inline src/sources/http.rs:62]: I think we should attempt to add the source type as a tag on the metric to match how we add it as a field to a log event.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Thanks so much for this PR @Xaeroxe!

I can see the motivation to send metrics over HTTP.

I am wondering if you are planning to use this to send events between two or more running `vector` instances though. Are you? If so, you may also want to consider the `vector` source and sink which already has support for metrics (encoding events as protobuf rather than JSON).

That being said, I think it'd be useful to be able to send metrics over HTTP regardless so this is definitely still relevant.

I think there are a few high-level issues to address before I go over this with a finer comb. I'm curious to get your thoughts on them and am happy to support throughout this process. We appreciate the contribution.

## Serialization / Deserialization

I think the serde serializing and deserializing logic you added is actually not needed given the already existing trait definitions:

```
impl TryFrom<serde_json::Value> for Event
impl TryInto<serde_json::Value> for Event
```

For example, to convert an `Event` from a `serde_json::Value, you can use `Event::try_from(value)` and, to a `serde_json::Value` (which can then be serialized as a string), you can use `let value: serde_json::Value = event.try_into().unwrap()` (you'd want to handle the error rather than panic'ing, but hopefully that gives you an idea).

I'd rather depend on those existing trait definitions than add additional serialization/deserialization routines as they are likely to not match. I realize this continues to limit us to JSON representations for events, but that should be fine for now. We can reapproach serializing `Event`s as different formats in the future if we need.

@hoverbear probably has thoughts here as she's been working on the event data model.

## Making the metrics mode optional

I think I'd actually prefer to just make this the default behavior for the `http` sink and source rather than an option people need to opt into.

## Text encoding

Rather than erroring if users attempt to use metrics with the `text` encoding for both the `http` source and sink, I think I'd prefer to just define a text format for metrics that can be sent and parsed.

We currently encode metrics using [Prometheus's text-based encoding](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md) for the `console` sink. I think this could make sense here too. We already have a parser for this encoding that we use with the `prometheus` source.

@binarylogic may have thoughts here.

**Comment by Xaeroxe** [comment]
> consider the vector source and sink

I did, I really do need to use HTTP for infrastructure reasons though.

> I think I'd actually prefer to just make this the default behavior for the http sink and source rather than an option people need to opt into.

Fine by me, making it opt in was done to improve compatibility with 0.10, but I certainly don't need such compatibility.

> Rather than erroring if users attempt to use metrics with the text encoding for both the http source and sink, I think I'd prefer to just define a text format for metrics that can be sent and parsed.

So long as there isn't ambiguity between metrics and logging in this configuration that's fine by me. I'll implement whatever format you want me to.

> I'd rather depend on those existing trait definitions than add additional serialization/deserialization routines as they are likely to not match. 

So, knowing now that backwards compatibility isn't a concern, why not just remove these trait implementations and use `serde_json::from_str::<Event>` instead? The serde data model gives you access to a lot of representations that could be useful for various wire formats, including JSON. Most of the logic I added was just to preserve the existing data format. I'm not especially tied to that data format if timber.io isn't tied to it.

File: src/sinks/http.rs (hunk 1 above)
Comment (Xaeroxe) [reply src/sinks/http.rs:154]: This is yet another shim for backwards compatibility purposes. It will likely be removed since you've indicated backwards compatibility isn't something we're concerned with.

**Comment by jszwedko** [comment]
Thanks for the responses!

> So long as there isn't ambiguity between metrics and logging in this configuration that's fine by me. I'll implement whatever format you want me to.

:+1: I'm partial to just using the prometheus format for text encoding given we already do that for the `console` sink. I'll let @binarylogic offer any alternate opinions though.

> So, knowing now that backwards compatibility isn't a concern, why not just remove these trait implementations and use serde_json::from_str::<Event> instead? The serde data model gives you access to a lot of representations that could be useful for various wire formats, including JSON. Most of the logic I added was just to preserve the existing data format. I'm not especially tied to that data format if timber.io isn't tied to it

Yeah, I guess I'm just feeling like there is a reason that was not previously done and instead the `TryInto` and `TryFrom` definitions were added instead. If amenable, I'd like to keep the scope of this PR to just the addition of metrics support for the `http` sink and source and we can break off general `serde` support into a separate issue. I know that @Hoverbear is actively working on refactoring the event model so she may have further thoughts. @lukesteensen may know more of the history of why this wasn't previously done.

**Comment by Xaeroxe** [comment]
I'd like it to be known that since you asked other maintainers for reviews that I do not intend to alter this PR until we hear from them, even though there are requested changes currently.

**Comment by Hoverbear** [comment]
@jszwedko So ideally any time a metric is serialized or deserialized it'd actually go through serde itself. So `LogEvent` for example has `#[derive(Serialize, Deserialize)]` which makes it able to be serialized or deserialized into any format. It has a lot of generic options (see https://serde.rs/container-attrs.html, for example).

I think the big question here is "How to represent it"

File: .meta/sinks/http.toml.erb [hunk 4]
```diff
@@ -103,6 +103,15 @@ type = "string"
 examples = ["https://10.22.212.22:9000/_health"]
 description = "A URI that Vector can request in order to determine the service health."
 
+[sinks.http.options.metrics_plus_mode]
```
Comment (Hoverbear) [inline .meta/sinks/http.toml.erb:106]: I'd prefer this to be removed. We already have an `encoding` option which supports json.

Comment (Hoverbear) [inline .meta/sinks/http.toml.erb:106]: I'd prefer this to be removed. We already have an `encoding` option which supports json.

File: .meta/sinks/http.toml.erb [hunk 5]
```diff
@@ -103,6 +103,15 @@ type = "string"
...
+type = "bool"
+examples = [true, false]
+description = """\
+When set to true this sink will use a text format that supports metrics. \
+The default format only supports logs. The receiving source must have the same value\
```
Comment (Hoverbear) [inline .meta/sinks/http.toml.erb:111]: Users can already transmit metrics between vector sources and sinks. This caveat kind of limits the use of this feature.

File: .meta/sources/http.toml.erb [hunk 6]
```diff
@@ -33,6 +33,15 @@ The expected encoding of received data. Note that for `json` and `ndjson` \
 encodings, the fields of the JSON objects are output as separate fields.\
 """
 
+[sources.http.options.metrics_plus_mode]
```
Comment (Hoverbear) [inline .meta/sources/http.toml.erb:36]: Please remove.

File: src/event/log_event.rs [hunk 7]
```diff
@@ -170,15 +171,6 @@ impl IntoIterator for LogEvent {
     }
 }
 
-impl Serialize for LogEvent {
```
Comment (Hoverbear) [inline src/event/log_event.rs:173]: Nice! This was on my list. :)

This is definitely outside of the scope of this PR though.

File: src/event/mod.rs [hunk 8]
```diff
@@ -92,6 +93,10 @@ fn timestamp_to_string(timestamp: &DateTime<Utc>) -> String {
     timestamp.to_rfc3339_opts(SecondsFormat::AutoSi, true)
 }
 
+fn string_to_timestamp(string: &str) -> Result<DateTime<Utc>, ParseError> {
```
Comment (Hoverbear) [inline src/event/mod.rs:96]: This appears to be dead code once other comments have been reflected.

File: src/event/value.rs [hunk 9]
```diff
@@ -1,41 +1,96 @@
...
-use serde::{Serialize, Serializer};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
 use std::collections::BTreeMap;
 use std::convert::{TryFrom, TryInto};
 use std::iter::FromIterator;
 use toml::value::Value as TomlValue;
 
-#[derive(PartialEq, Debug, Clone, is_enum_variant)]
+#[derive(PartialEq, Debug, Clone, is_enum_variant, Deserialize, Serialize)]
+#[serde(untagged)]
```
Comment (Hoverbear) [inline src/event/value.rs:12]: This looks out of scope of this PR!

(It's a heckin' good idea though and do feel welcome to make a separate PR for this!)

File: src/event/value.rs [hunk 10]
```diff
@@ -1,41 +1,96 @@
...
+        serialize_with = "serialize_as_string_timestamp",
+        deserialize_with = "deserialize_as_string_timestamp"
+    )]
+    Timestamp(DateTime<Utc>),
+    #[serde(
+        serialize_with = "serialize_as_string_bytes",
+        deserialize_with = "deserialize_as_string_bytes"
+    )]
+    Bytes(Bytes),
+    #[serde(
```
Comment (Hoverbear) [inline src/event/value.rs:32]: Similarly, I don't think this PR is the right place to be making changes to our core data model deserialization and this is going to impact existing work in this area.

**Review by Hoverbear — CHANGES_REQUESTED** [review]
I don't really understand what's happening here. 😅 

* Why does this PR not allow the HTTP source and sink to injest JSON metrics which have already defined serialization and deserialization? I'm not so keen on non-standard formats
* Why does it seem to require some special text format that we have no RFC or spec for? Vector Source/sinks exist and can already do this.
* The HTTP sink and source have other components depending on them, does this also enable support in those?

I'd really love to merge this PR to add support to the HTTP source/sink for sending/receiving JSON metrics, but I am cautious of it supporting much more.

In regards to code quality etc, great job, it looks great.

**Comment by binarylogic** [comment]
Hi @Xaeroxe thanks a lot for your contribution. We appreciate your willingness to dive in and write code to solve your problem.

I hate to say this, but I think we need to rethink the approach here. This is a case where some consensus upfront would have helped guide development. Our [contribution guide](https://github.com/timberio/vector/blob/master/CONTRIBUTING.md#your-first-contribution) notes finding or opening an issue to obtain consensus on a direction. For example, #3996 proposes supporting more formats (codecs) in the `http` source and that would likely be the correct solution here.

We'd love to see you continue working on this. If you'd like to give it another shot, why don't you chime in on #3996 with a proposed direction?

Closing this in favor for that.

**Comment by Xaeroxe** [comment]
That's understandable. I'd like to be known there were a few misunderstandings here. @Hoverbear commented several times that the data model was being altered. Special care was taken to make sure the serialized data, and deserialization code, had identical behaviour. The only thing that changed was that this PR used macros to replace a lot of the handwritten code. Additionally, deserialization code for this data format was added.

The only change made to the data model, is that when `metrics_plus_mode` was activated the JSON serialization format was extended so that metrics and logs could be differentiated. This extension was a superset of the original data model.

I'm not opposed to https://github.com/timberio/vector/issues/3907 in the long term, though I do think as a more short term goal this PR was more workable than perhaps it initially appeared.

**Comment by binarylogic** [comment]
Yeah, I'm concerned that the changes here would make it more difficult to implement the proper solution down the line, which is to decouple our metrics parsers (statsd, prometheus, and influx) into reusable codecs that could be enabled for the `http` source. This is something we are currently addressing, so I don't see this solution being that far off.

**Comment by Hoverbear** [comment]
@Xaeroxe sorry, yes, you're right. You did not change the data model here. :)

**Comment by Xaeroxe** [comment]
> Yeah, I'm concerned that the changes here would make it more difficult to implement the proper solution down the line

@binarylogic 

My understanding is that for the final solution we want to be able to feed metric and log structures through a variety of possible serializers and deserializers, and that those data formats should be decoupled from the means of transmission. The existence of this data model doesn't impede other ones from being added. When the time is right, timberio could delete these additions if this data model ends up becoming dead code.

I also believe these changes might actually help you with that goal. The serde data format is a format neutral abstract representation. This PR enhances the existing serde interface by adding a few implementations of `serde::Deserialize`. For some of your target data formats, you might be able to implement them as `serde::Serializer` and `serde::Deserializer`, or even leverage existing serializers and deserializers. This would give you robust code for serialization and deserialization that isn't as prone to common mistakes with handwritten serializers.

**Comment by Xaeroxe** [comment]
`metrics_plus_mode` has been removed from my branch, the code now always assumes this value is true. Unfortunately that won't reflect here because the PR has been closed.

**Comment by Xaeroxe** [comment]
If you'd like to review those changes you can look at them here: https://github.com/vivint-smarthome/vector/compare/master...vivint-smarthome:http-metrics?expand=1

**Comment by Hoverbear** [comment]
Yay it's back!

**Comment by Xaeroxe** [comment]
That's a lot of merge conflicts. I'll get them worked out.

File: src/sources/http.rs [hunk 11]
```diff
@@ -366,8 +372,8 @@ mod tests {
 
         let (rx, addr) = source(Encoding::Json, vec![]).await;
 
-        assert_eq!(200, send(addr, r#"[{"key":"value"}]"#).await);
-        assert_eq!(200, send(addr, r#"{"key2":"value2"}"#).await);
+        assert_eq!(200, send(addr, r#"[{"Log":{"key":"value"}}]"#).await);
+        assert_eq!(200, send(addr, r#"{"Log":{"key2":"value2"}}"#).await);
```
Comment (Hoverbear) [inline src/sources/http.rs:376]: So this seems like it'd be a breaking change.

Comment (Xaeroxe) [reply src/sources/http.rs:376]: It would yeah. That's why the `metrics_plus_mode` existed to begin with.

Comment (Hoverbear) [reply src/sources/http.rs:376]: Yeah, then I almost lean towards punting this until we have some greater RFC/architecture work for specialized handling of this kind of thing. I'm sorry. :(

Does the `vector` source/sink not work for you?

Comment (Xaeroxe) [reply src/sources/http.rs:376]: Unfortunately for my use case I really do need `http`. Infrastructure reasons.

I suppose I can just run my own fork until these discussions happen.

**Comment by jamtur01** [comment]
As an FYI - I re-opened so @Xaeroxe could tidy up the branch rather than suggest we're going to merge until we've had further discussions! :)

**Comment by lukesteensen** [comment]
> My understanding is that for the final solution we want to be able to feed metric and log structures through a variety of possible serializers and deserializers, and that those data formats should be decoupled from the means of transmission.

This is roughly correct, but the intent is to have a variety of parsers and encoders rather than serializers and deserializers. The difference is a bit subtle, but mostly we don't want to couple the shape our internal data model to these external APIs. They'll be intended to translate from existing external models into our internal model, which could change at any time without affecting those interfaces.

**Comment by Xaeroxe** [comment]
There doesn't seem to be a well agreed upon path forward at this point, and this PR is rapidly generating merge conflicts. So I think it's appropriate to close this and revisit it at another date. I'm keeping the branch for my own purposes though.
